File: projects/robots/telerob/telemax/controllers/telemax_keyboard_controller/telemax_keyboard_controller.py
# Copyright 1996-2018 Cyberbotics Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""telemax_keyboard_controller controller."""

from controller import Keyboard, Robot

robot = Robot()
timestep = int(robot.getBasicTimeStep())

keyboard = robot.getKeyboard()
keyboard.enable(timestep)

frontMotor = robot.getMotor('front motor')
rearMotor = robot.getMotor('rear motor')
frontLeftTrack = robot.getMotor('left front track')
frontRightTrack = robot.getMotor('right front track')
rearLeftTrack = robot.getMotor('left rear track')
rearRightTrack = robot.getMotor('right rear track')


frontLeftTrack.setPosition(float('inf'))
frontRightTrack.setPosition(float('inf'))
rearLeftTrack.setPosition(float('inf'))
rearRightTrack.setPosition(float('inf'))
frontLeftTrack.setVelocity(0)
frontRightTrack.setVelocity(0)
rearLeftTrack.setVelocity(0)
rearRightTrack.setVelocity(0)

centralLed = robot.getLED('central led')
leftLed = robot.getLED('left led')
rightLed = robot.getLED('right led')

position = [0.0, 0.0]
speed = 0.0
angle = 0.0
previousKeys = []

while robot.step(timestep) != -1:
    key = keyboard.getKey()
    currentKeys = []
    while key != -1:
        currentKeys.append(key)
        if key == ord('Q'):
            position[0] += 0.01
        elif key == ord('A'):
            position[0] -= 0.01
        elif key == ord('W'):
            position[1] += 0.01
        elif key == ord('S'):
            position[1] -= 0.01
        elif key == Keyboard.UP:
            speed += 0.005
        elif key == Keyboard.DOWN:
            speed -= 0.005
        elif key == Keyboard.RIGHT:
            angle += 0.002
        elif key == Keyboard.LEFT:
            angle -= 0.002
        elif key == ord('I') and key not in previousKeys:
            leftLed.set(not leftLed.get())
        elif key == ord('O') and key not in previousKeys:
            centralLed.set(not centralLed.get())
        elif key == ord('P') and key not in previousKeys:
            rightLed.set(not rightLed.get())
        key = keyboard.getKey()

    previousKeys = currentKeys
    frontMotor.setPosition(position[0])
    rearMotor.setPosition(position[1])
    frontLeftTrack.setVelocity(speed + angle)
    frontRightTrack.setVelocity(speed - angle)
    rearLeftTrack.setVelocity(speed - angle)
    rearRightTrack.setVelocity(speed + angle)

File: projects/robots/telerob/telemax/controllers/telemax_keyboard_controller/telemax_keyboard_controller.c
/*
 * Copyright 1996-2019 Cyberbotics Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <webots/camera.h>
#include <webots/keyboard.h>
#include <webots/led.h>
#include <webots/motor.h>
#include <webots/robot.h>

#define TIME_STEP 32

int main(int argc, char **argv) {
  wb_robot_init();

  wb_keyboard_enable(TIME_STEP);

  WbDeviceTag front_motor = wb_robot_get_device("front motor");
  WbDeviceTag rear_motor = wb_robot_get_device("rear motor");
  WbDeviceTag front_left_track = wb_robot_get_device("left front track");
  WbDeviceTag front_right_track = wb_robot_get_device("right front track");
  WbDeviceTag rear_left_track = wb_robot_get_device("left rear track");
  WbDeviceTag rear_right_track = wb_robot_get_device("right rear track");

  wb_motor_set_position(front_left_track, INFINITY);
  wb_motor_set_position(front_right_track, INFINITY);
  wb_motor_set_position(rear_left_track, INFINITY);
  wb_motor_set_position(rear_right_track, INFINITY);
  wb_motor_set_velocity(front_left_track, 0.0);
  wb_motor_set_velocity(front_right_track, 0.0);
  wb_motor_set_velocity(rear_left_track, 0.0);
  wb_motor_set_velocity(rear_right_track, 0.0);

  WbDeviceTag central_led = wb_robot_get_device("central led");
  WbDeviceTag left_led = wb_robot_get_device("left led");
  WbDeviceTag right_led = wb_robot_get_device("right led");

  WbDeviceTag camera = wb_robot_get_device("gripper camera");

  double position[2] = {0.0, 0.0};
  double speed = 0.0;
  double angle = 0.0;
  bool led_key_pressed[3] = {false, false, false};
  bool camera_key_pressed = false;

  while (wb_robot_step(TIME_STEP) != -1) {
    int key = wb_keyboard_get_key();
    bool current_led_key_pressed[3] = {false, false, false};
    bool current_camera_key_pressed = false;
    while (key != -1) {
      switch (key) {
        case WB_KEYBOARD_HOME:
          position[0] += 0.01;
          break;
        case WB_KEYBOARD_END:
          position[0] -= 0.01;
          break;
        case WB_KEYBOARD_PAGEUP:
          position[1] += 0.01;
          break;
        case WB_KEYBOARD_PAGEDOWN:
          position[1] -= 0.01;
          break;
        case WB_KEYBOARD_UP:
          speed -= 0.005;
          break;
        case WB_KEYBOARD_DOWN:
          speed += 0.005;
          break;
        case WB_KEYBOARD_RIGHT:
          angle -= 0.002;
          break;
        case WB_KEYBOARD_LEFT:
          angle += 0.002;
          break;
        case 'I':
          current_led_key_pressed[0] = true;
          if (!led_key_pressed[0])
            wb_led_set(left_led, !wb_led_get(left_led));
          break;
        case 'O':
          current_led_key_pressed[1] = true;
          if (!led_key_pressed[1])
            wb_led_set(central_led, !wb_led_get(central_led));
          break;
        case 'P':
          current_led_key_pressed[2] = true;
          if (!led_key_pressed[2])
            wb_led_set(right_led, !wb_led_get(right_led));
          break;
        case 'C':
          current_camera_key_pressed = true;
          if (!camera_key_pressed) {
            if (wb_camera_get_sampling_period(camera) > 0)
              wb_camera_disable(camera);
            else
              wb_camera_enable(camera, 32);
          }
          break;
      }
      key = wb_keyboard_get_key();
    }

    led_key_pressed[0] = current_led_key_pressed[0];
    led_key_pressed[1] = current_led_key_pressed[1];
    led_key_pressed[2] = current_led_key_pressed[2];
    camera_key_pressed = current_camera_key_pressed;

    wb_motor_set_position(front_motor, position[0]);
    wb_motor_set_position(rear_motor, position[1]);
    wb_motor_set_velocity(front_left_track, speed + angle);
    wb_motor_set_velocity(front_right_track, speed - angle);
    wb_motor_set_velocity(rear_left_track, -speed - angle);
    wb_motor_set_velocity(rear_right_track, -speed + angle);
  };

  wb_robot_cleanup();

  return 0;
}
